subsys/net/lib/http_server/src/server.c

```diff
@@ -0,0 +1,538 @@
+}
+
+int http2_server_init(struct http2_server_ctx *ctx,
+		      struct http2_server_config *config)
+{
+    /* Create a socket */
+    ctx->sockfd = socket(config->address_family, SOCK_STREAM, 0);
+    if (ctx->sockfd < 0) {
+	perror("socket");
+	exit(EXIT_FAILURE);
+    }
+
+    /* Set up the server address struct according to address family */
+    if (config->address_family == AF_INET) {
+	struct sockaddr_in serv_addr;
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sin_family = AF_INET;
+	serv_addr.sin_addr.s_addr = INADDR_ANY;
+	serv_addr.sin_port = htons(config->port);
+
+	if (bind(ctx->sockfd, (struct sockaddr *)&serv_addr,
+		 sizeof(serv_addr)) < 0) {
+	    perror("bind");
+	    exit(EXIT_FAILURE);
+	}
+    } else if (config->address_family == AF_INET6) {
+	struct sockaddr_in6 serv_addr;
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sin6_family = AF_INET6;
+	serv_addr.sin6_addr = in6addr_any;
+	serv_addr.sin6_port = htons(config->port);
+
+	if (bind(ctx->sockfd, (struct sockaddr *)&serv_addr,
+		 sizeof(serv_addr)) < 0) {
+	    perror("bind");
+	    exit(EXIT_FAILURE);
+	}
+    }
+
+    /* Listen for connections */
+    if (listen(ctx->sockfd, MAX_CLIENTS) < 0) {
+	perror("listen");
+	exit(EXIT_FAILURE);
+    }
+
+    /* Initialize client_fds */
+    memset(ctx->client_fds, 0, sizeof(ctx->client_fds));
+    ctx->client_fds[0].fd = ctx->sockfd;
+    ctx->client_fds[0].events = POLLIN;
+    ctx->num_clients = 0;
+
+    return ctx->sockfd;
+}
+
+int http2_server_start(int server_fd, struct http2_server_ctx *ctx)
+{
+    printf("Press 'Ctrl+C' to quit\n");
+    printf("Waiting for incoming connections on port 8080...\n");
+
+    while (1) {
+	int ret = poll(ctx->client_fds, ctx->num_clients + 1, -1);
+	if (ret < 0) {
+	    perror("poll failed");
+	    exit(EXIT_FAILURE);
+	}
+
+	for (int i = 0; i <= ctx->num_clients; i++) {
+	    if (ctx->client_fds[i].revents & POLLIN) {
+		if (i == 0) {
+		    int new_socket = accept_new_client(server_fd);
+
+		    for (int j = 1; j < MAX_CLIENTS; j++) {
+			if (ctx->client_fds[j].fd == 0) {
+			    ctx->client_fds[j].fd = new_socket;
+			    ctx->client_fds[j].events = POLLIN;
+
+			    if (j > ctx->num_clients) {
+				ctx->num_clients = j;
+			    }
+			    break;
+			}
+		    }
+		} else {
+		    /* Read the client data */
+		    int valread =
+			recv(ctx->client_fds[i].fd, buffer, BUFFER_SIZE - 1, 0);
+		    printf("valread length: %u\n", valread);
+		    if (valread < 0) {
+			perror("ERROR reading from "
+			       "socket");
+			close_client_connection(ctx, i);
+		    } else if (valread == 0) {
+			printf("Connection closed by "
+			       "peer.\n");
+			close_client_connection(ctx, i);
+		    } else {
+			buffer[valread] = '\0';
+			printf("buffer %s\n", buffer);
+			http_parser_init(&parser, HTTP_REQUEST);
+			http_parser_settings_init(&parserSettings);
+			parserSettings.on_header_field = on_header_field;
+			http_parser_execute(&parser, &parserSettings, buffer,
```

Comment:
Shouldn't the HTTP1 parser be executed only after we identify we deal with HTTP1 request?

Reply:
I called here because I needed to see if there is an upgrade: h2c in the header field.

Reply:
The thing is - you cannot feed the data to the HTTP1 parser unless you're certain it's a HTTP1 request. Otherwise, you may get unexpected results. So the order of action should be:
* Read data on a socket until you have `strlen(preface) bytes`
* If preface matches HTTP2 request - proceed as with HTTP2 request.
* Otherwise, proceed with HTTP1 request. Only now you can feed that data to the HTTP1 parser, and search for the upgrade header. If there's a header, reply and await next request (HTTP2 this time).